Mathlib/Analysis/Calculus/MeanValue.lean

```diff
@@ -596,6 +596,52 @@ theorem eqOn_of_fderivWithin_eq (hs : Convex ℝ s) (hf : DifferentiableOn 𝕜
   refine hs.is_const_of_fderivWithin_eq_zero (hf.sub hg) (fun z hz => ?_) hx hy
   rw [fderivWithin_sub (hs' _ hz) (hf _ hz) (hg _ hz), sub_eq_zero, hf' _ hz]
 
+/-- If `f` has zero derivative on an open set, then `f` is locally constant on `s`. -/
+theorem _root_.IsOpen.isOpen_inter_preimage_of_fderiv_eq_zero
+    (hs : IsOpen s) (hf : DifferentiableOn 𝕜 f s)
+    (hf' : ∀ x ∈ s, fderiv 𝕜 f x = 0) (t : Set G) : IsOpen (s ∩ f ⁻¹' t) := by
+  refine Metric.isOpen_iff.mpr fun y ⟨hy, hy'⟩ ↦ ?_
+  obtain ⟨r, hr, h⟩ := Metric.isOpen_iff.mp hs y hy
+  refine ⟨r, hr, Set.subset_inter h fun x hx ↦ ?_⟩
+  have := (convex_ball y r).is_const_of_fderivWithin_eq_zero (hf.mono h) ?_ hx (mem_ball_self hr)
+  · simpa [this]
+  · intro z hz
+    simpa only [fderivWithin_of_isOpen Metric.isOpen_ball hz] using hf' z (h hz)
+
+/-- If `f` has zero derivative on a connected open set, then `f` is constant on `s`. -/
+theorem _root_.IsOpen.is_const_of_fderiv_eq_zero
+    (hs : IsOpen s) (hs' : IsPreconnected s) (hf : DifferentiableOn 𝕜 f s)
+    (hf' : ∀ x ∈ s, fderiv 𝕜 f x = 0) : ∃ a, ∀ x ∈ s, f x = a := by
```

Comment:
Could you provide a version with the conclusion `∀ x ∈ s, ∀ y ∈ s, f x = f y`? That could be nicer in certain contexts so that you don't have to extract the witness from the existential.